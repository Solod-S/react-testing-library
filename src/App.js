import React, { useState, useEffect } from "react";
import logo from "./logo.svg";
import "./App.css";

/*

Search variants:

  getBy:                    queryby:                    findBy:

- getByText               - queryByText               - findByText
- getByRole               - queryByRole               - findByRole
- getByLabelText          - queryByLabelText          - findByLabelText
- getByPlaceholderText    - queryByPlaceholderText    - findByPlaceholderText
- getByAltText            - queryByAltText            - findByAltText
- getByDisplayValue       - queryByDisplayValue       - findByDisplayValue
- getAllBy                - queryAllBy                - findAllBy
*/

/*
Assertive Functions:

- toBeDisabled            - toBeEnabled               - toBeEmpty
- toBeEmptyDOMElement     - toBeInTheDocument         - toBeInvalid
- toBeRequired            - toBeValid                 - toBeVisible
- toContainElement        - toContainHTML             - toHaveAttribute
- toHaveClass             - toHaveFocus               - toHaveFormValues
- toHaveStyle             - toHaveTextContent         - toHaveValue
- toHaveDisplayValue      - toBeChecked               - toBePartiallyChecked
- toHaveDescription
*/

const getUser = () => Promise.resolve({ id: 1, name: "Sergey" });

const Search = ({ value, onChange, children }) => (
  <div>
    <label htmlFor="search">{children}</label>
    <input required placeholder="search text..." id="search" type="text" value={value} onChange={onChange} />
  </div>
);

const App = () => {
  const [search, setSearch] = useState("");
  const [user, setUser] = useState();

  useEffect(() => {
    const loadUser = async () => {
      const user = await getUser();
      setUser(user);
    };
    loadUser();
  }, []);

  const handleChange = ({ target }) => {
    setSearch(target.value);
  };

  return (
    <div>
      {user && <h2>Logged in as {user.name}</h2>}
      <img className="image" src="" alt="search image" />
      <Search value={search} onChange={handleChange}>
        Search:
      </Search>
      <p>Searches for {search ? search : "..."}</p>
    </div>
  );
};

export default App;
