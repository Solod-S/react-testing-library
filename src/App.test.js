import { render, screen } from "@testing-library/react";
import App from "./App";

// очень часто наша задачей будет найти поле ввода и проверить результат

test("renders learn react link", () => {
  render(<App />);
  const linkElement = screen.getByText(/Search:/i);
  expect(linkElement).toBeInTheDocument();
});
// test("renders learn react link 2", () => {
//   const { getByText } = render(<App />);
//   const linkElement = getByText(/Search:/i);
//   expect(linkElement).toBeInTheDocument();
// });
test("screen", () => {
  render(<App />);
  screen.debug();
  // рендер всей разметки в консоли
});

test("screen 2", () => {
  const { asFragment } = render(<App />);
  expect(asFragment(<App />)).toMatchSnapshot();
  // снимок
});

describe("App element sect basics", () => {
  it("render App component ", () => {
    render(<App />);
    // рендер компонента
    screen.debug();
    // рендер всей разметки в консоли

    expect(screen.getByText(/search:/i)).toBeInTheDocument();
    // проверяем наличие текста серч (через регулярку не такое точное сравнение, /i - не чувствительность к регистру)

    expect(screen.getByRole("textbox")).toBeInTheDocument();
    // выбор элемента по лейблу

    expect(screen.getByLabelText(/search:/i)).toBeInTheDocument();
    // выбор элемента по его тексту в лейбле

    expect(screen.getByPlaceholderText("search text...")).toBeInTheDocument();
    // выбор элемента по тексту плейсхолдера

    expect(screen.getByAltText("search image")).toBeInTheDocument();
    // выбор элемента по альтернативному тексту

    expect(screen.getByDisplayValue("")).toBeInTheDocument();
    // поиск элемента по отображаемому значению
  });
});

describe("App element sect basics part 2", () => {
  it("queryByText", () => {
    render(<App />);
    // рендер компонента
    screen.debug();
    // рендер всей разметки в консоли

    expect(screen.queryByText(/Searches for React/i)).toBeNull();
    // проверяем элемент которого еще нет в разметке ("Searches for" есть, а вот "React" мы добавим из инпута)
    // и говорим что изначально в документе ее нету
  });
});

describe("Async handle element", () => {
  // изначально элемента нету но после выполнения асинхроного кода он появиться
  it("queryByText", async () => {
    render(<App />);
    // рендер компонента
    screen.debug();
    // рендер всей разметки в консоли
    expect(screen.queryByText(/Logged in as/i)).toBeNull();
    // ищу строку в документа и говорю что ее нету
    expect(await screen.findByText(/Logged in as/i)).toBeInTheDocument();
    // далее когда выполниться асинхроный код появиться эта строка

    expect(screen.getByAltText(/search image/i)).toHaveClass("image");
    // находим по альтернативному тексту + проверяем класс
    expect(screen.getByPlaceholderText(/search text.../i)).toBeRequired();
    //  проверяем на .toBeRequired(); если нужно чтобы отрицательное .not.toBeRequired();
    expect(screen.getByPlaceholderText(/search text.../i)).toBeEmpty();
    // проверяем что инпут пустой
    expect(screen.getByPlaceholderText(/search text.../i)).toHaveAttribute("id");
    // проверяем айди
    screen.debug();
  });
});

// нужно найти элемент ===getBy
// нужно показать что элемента нет в доме ===queryBy
// нужно показать что элемента изначально нет но потом он появиться ===findBy
// для нескольких элементов ===AllBy

// 28
