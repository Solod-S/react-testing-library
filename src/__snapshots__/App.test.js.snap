// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`screen 2 1`] = `
<DocumentFragment>
  <div>
    <img
      alt="search image"
      class="image"
      src=""
    />
    <div>
      <label
        for="search"
      >
        Search:
      </label>
      <input
        id="search"
        placeholder="search text..."
        required=""
        type="text"
        value=""
      />
    </div>
    <p>
      Searches for ...
    </p>
  </div>
</DocumentFragment>
`;
